### DISCUSSION TITLE
Add organization / project / user / API key management API

### DISCUSSION THREAD

**Discussion by EricAntidot** [langfuse/langfuse#3243] [Ideas]
### Describe the feature or potential improvement

We are self hosting a couple of langfuse instance in our infrastructure and we want to be able to scripts (or integrate with other app) the management part of langfuse project.

Namely I want to be able to 

- create / delete / update an organization
- create / delete / update a project
- create / delete / update a user
- create / delete / update an API Key

i.e. we currently do that using langfuse UI but it would be more efficient to do it programmatically.

The recent added capability to do part of this during init:  https://langfuse.com/changelog/2024-09-04-headless-initialization-of-self-hosted-deployments is a step forward. But we want more ;-)

**Comment by EricAntidot** [comment]
Related issue/discussions:
https://github.com/langfuse/langfuse/issues/517

**Comment by marcklingen** [comment]
thanks for creating this thread, agree that an admin api qould be useful, either on the instance and/or on the organizational level

let's collect use cases and track overall interest in this thread

**Comment by thiagolealassis** [comment]
I agree!

**Comment by co3k** [comment]
I have been wanting to manage multiple Langfuse projects programmatically.

I noticed that the ability to delete API keys has been added in https://github.com/langfuse/langfuse/pull/3448. This new feature uses the ADMIN_API_KEY to perform the deletions.
In a similar way, creating CRUD APIs for API keys and projects seems feasible and would also meet my needs. I am considering making such a contribution.

By the way, if we want to manage organizations and projects across the board, it seems the most straightforward approach is to enable all relevant APIs to be executed directly via the ADMIN_API_KEY. Although at first glance this approach may seem risky, if the ADMIN_API_KEY can control project-specific API keys (which are currently necessary for executing various APIs), the actual risk does not seem to increase. Therefore, as a first step, it seems reasonable to expand the scope of the ADMIN_API_KEY.

What do you think?

**Reply by marcklingen** [reply]
This admin api is not yet finalized (might change at any time) and was added to manage some complexities in running Langfuse Cloud.

We plan to add a proper admin api to manage both
- organizations
  - via organization-level api keys (to be introduced)
  - Example actions: CRUD projects, add/remove members, creates/delete api keys)
- langfuse self-hosted instances
  - `ADMIN_API_KEY` to be set via environment
  - Example actions: CRUD orgs, create/delete org api keys

**Reply by EricAntidot** [reply]
Do you have any timeline on those feature?
I'll be glad to review of contribute to the design of this API.

**Reply by co3k** [reply]
@marcklingen 

This sounds like a very promising direction!

For the initial implementation of the organization-level API key issuance functionality that you are envisioning, would it be reasonable to assume that it should be roughly equivalent to the current project-level API key issuance functionality?
If so, would it be alright if I propose an implementation via a pull request?

Additionally, considering compatibility with the SDK, I believe it would be best to distinguish between project-level and organization-level API keys solely through their prefixes (for instance, project-level API keys currently use the "pk-" prefix, so organization-level API keys could use the "ok-" prefix). How do you envision distinguishing between these key types?

**Reply by marcklingen** [reply]
> Do you have any timeline on those feature?
> I'll be glad to review of contribute to the design of this API.

Not yet as there a couple of high-prio changes that we will address beforehand. I think scope/design of apis is quite clear. Please share any additional use cases in addition to the ones mentioned above that you'd need.

> For the initial implementation of the organization-level API key issuance functionality that you are envisioning, would it be reasonable to assume that it should be roughly equivalent to the current project-level API key issuance functionality?
If so, would it be alright if I propose an implementation via a pull request?

I think this needs a larger revamp of api keys as it would be nice to also use these organization level api keys to perform project-level actions (e.g. use a single api key to add traces to multiple projects, to be differentiated at run-time).

> Additionally, considering compatibility with the SDK, I believe it would be best to distinguish between project-level and organization-level API keys solely through their prefixes (for instance, project-level API keys currently use the "pk-" prefix, so organization-level API keys could use the "ok-" prefix). How do you envision distinguishing between these key types?

Good question. I think the ideal scenario here is to only have org-level api keys with custom scopes and migrate the existing keys to have scopes for the single project only. Otherwise this change would add a lot of complexity to the project that is not easy to understand/maintain.

**Comment by marcklingen** [comment]
I've just copied a lot of the context from this thread to the previously existing thread (with more upvotes). Please add any additional thoughts and your upvotes there. Thank you!

https://github.com/orgs/langfuse/discussions/1007
